services/src/main/java/org/fao/geonet/services/metadata/BatchOpsMetadataReindexer.java

```diff
@@ -41,111 +41,170 @@
+        probeName = new ObjectName(String.format("geonetwork:name=indexing-task,idx=%s", this.hashCode()));
+        exporter.registerManagedResource(this, probeName);
+        return processAsync(runInCurrentThread);
+    }
+
+    private String processAsync(boolean runInCurrentThread) throws Exception {
         int threadCount = ThreadUtils.getNumberOfThreads();
-        ExecutorService executor = null;
-        try {
+
             if (runInCurrentThread) {
                 executor = MoreExecutors.sameThreadExecutor();
             } else {
                 executor = Executors.newFixedThreadPool(threadCount);
             }
-            int[] ids = new int[metadata.size()];
-            int i = 0;
-            for (Integer id : metadata) ids[i++] = id;
+
+            int[] ids = metadata.stream().mapToInt(i -> i).toArray();
 
             int perThread;
             if (ids.length < threadCount) perThread = ids.length;
             else perThread = ids.length / threadCount;
+
             int index = 0;
 
             List<BatchOpsCallable> jobs = Lists.newArrayList();
             while (index < ids.length) {
                 int start = index;
                 int count = Math.min(perThread, ids.length - start);
                 // create threads to process this chunk of ids
-                jobs.add(createCallable(ids, start, count));
+                BatchOpsCallable task = new BatchOpsCallable(ids, start, count);
+                jobs.add(task);
 
                 index += count;
             }
-            List<Future<Void>> submitList = Lists.newArrayList();
-            for (i = 1; i < jobs.size(); i++) {
-                Future<Void> submit = executor.submit(jobs.get(i));
-                submitList.add(submit);
-            }
 
-            if (jobs.size() > 0) {
-                jobs.get(0).call();
-                for (Future<Void> future : submitList) {
-                    try {
-                        future.get();
-                    } catch (InterruptedException | ExecutionException e) {
-                        e.printStackTrace();
-                    }
-                }
+            List<CompletableFuture> submitList = Lists.newArrayList();
+            for (BatchOpsCallable job : jobs) {
+                CompletableFuture completed = CompletableFuture.runAsync(job, executor);
+                submitList.add(completed);
             }
-        } finally {
-            if (executor != null) {
-                executor.shutdown();
-            }
-        }
+
+            allCompleted = CompletableFuture.allOf(submitList.toArray(new CompletableFuture[submitList.size()]));
+            allCompleted.thenRun(this);
+            return probeName.toString();
     }
 
-    protected BatchOpsCallable createCallable(int[] ids, int start, int count) {
-        return new BatchOpsCallable(ids, start, count, getDataManager());
+    @Override
+    public void run() {
+        executor.shutdown();
```

Comment:
@cmangeat, can you explain a bit this code? It seems cleaning up stuff, but can't really see that the task is executed or scheduled here, unless is done automatically when doing `PROBE_CACHE.put(probeName, probeName);`?

Reply:
Sorry I see is used in `processAsync` to cleanup indeed, a bit odd the usage in my opinion, but ok, it's clear now.